### ISSUE TITLE
Drive redirection: fuse not umounted at logout

### ISSUE THREAD

**Issue by moobyfr** [neutrinolabs/xrdp#780]
Using latest xrdp with xorg driver, Microsoft Remote Desktop on Mac:
Il a share is set (drive redirection), the share is working, but at logout it isn't umounted on the server.
(still displayed in "mount" output)

**Comment by mightyraven** [comment]
I think you have the same problem (https://github.com/neutrinolabs/xrdp/issues/761).
I noticed if the client (Linux xfreerdp), there is no problem with umount.
Problem with only Microsoft Remote Desktop.

**Comment by metalefty** [comment]
Does the issue occur only at logout? What about disconnection keeping remote session?

**Comment by mightyraven** [comment]
> Does the issue occur only at logout? What about disconnection keeping remote session?

Yes. When disconnection keeping remote session, all everything works perfectly.

**Comment by metalefty** [comment]
@moobyfr and for you, too?

**Comment by moobyfr** [comment]
sometime, with xfreerdp, at end of the session, the fuse is umounted. but not always. For the disconnect and reconnect, the fusemount is still mounted

**Comment by mightyraven** [comment]
For more information on the issue:

When I log in under root on the server (locally, without the use of xrdp), and then connect remotely under the user using xrdp, problems with drive redirection does not occur. Disconnect, then reconnect. Everything is working fine. (Files: .xorgxrdp.11.log .xorgxrdp.12.log .xorgxrdp.13.log are not created. It creates only 2 files:.xorgxrdp.10.log .xorgxrdp.10.log.old)

But as soon as I log off root (locally, without the use of xrdp), disconnect and reconnect the drive redirection is broken. ( I have created the file: .xorgxrdp.10.log .xorgxrdp.11.log .xorgxrdp.12.log .xorgxrdp.13.log, and so on.)

**Comment by moobyfr** [comment]
I've managed to debug a little, there was already some issue with systemd about gpg-agent not exiting on session logout and bug #778

**Comment by DRN88** [comment]
Very simple workaround is the following:
`fusermount -u "$HOME/redirected_drives"`

My custom `startwm.sh` looks like this:

```bash
[root@lxcentos7test01 ~]# cat /usr/libexec/xrdp/startwm-mate-session.sh 
#!/bin/sh
# Ansible managed
mate-session
fusermount -u "$HOME/redirected_drives"
exit 1
```

Set drive folder name manually:
```bash
[root@lxcentos7test01 ~]# grep FuseMountName /etc/xrdp/sesman.ini 
FuseMountName=redirected_drives
```

**Comment by moobyfr** [comment]
I've made some more tests.
if I disable drive redirection on server side, only activating clipboard makes the use of fuse, and doesn't umount at exit.

(I've tested with xfreerdp)

**Comment by moobyfr** [comment]
Current test case:
* xrdp/xorg/xrdp compiled from sources (devel) with --enable-fuse 
* rdesktop 1.8.3 (from brew) without any parameters (except login/pass/server

Sometimes there is no call for `xfuse_deinit()` when logging out of the desktop

**Comment by metalefty** [comment]
Hmm, I also think `xfuse_deinit()` is missing somewhere. Need to find out.

**Comment by hzhu212** [comment]
@metalefty It's a little different for my case.

Each time I disconnect and reconnect to remote desktop, my remote dirves will duplicate. Here is the situation after two times of reconnection:

![2](https://user-images.githubusercontent.com/19285447/41111781-a330c8d2-6aaf-11e8-8ba8-6fa9ee3f1260.png)

It seems that the remote drives are pointing to the same location:

```sh
# ll -i thinclient_drives/
total 0
3 drwxr-xr-x. 1 root root 0 Jun  7 22:45 D:
3 drwxr-xr-x. 1 root root 0 Jun  7 22:45 D:
3 drwxr-xr-x. 1 root root 0 Jun  7 22:45 D:
```

My environment:

```sh
# cat /etc/redhat-release
CentOS Linux release 7.5.1804 (Core)
```
```sh
# rpm -qa xfce* | sort
xfce4-appfinder-4.12.0-4.el7.x86_64
xfce4-cpugraph-plugin-1.0.5-8.el7.x86_64
xfce4-genmon-plugin-3.4.0-8.el7.x86_64
xfce4-panel-4.12.1-4.el7.x86_64
xfce4-power-manager-1.6.0-2.el7.x86_64
xfce4-pulseaudio-plugin-0.2.5-2.el7.x86_64
xfce4-session-4.12.1-8.el7.x86_64
xfce4-session-engines-4.12.1-8.el7.x86_64
xfce4-settings-4.12.1-1.el7.x86_64
xfce4-terminal-0.6.3-4.el7.x86_64
xfce-polkit-0.2-8.el7.x86_64
```

A little strange, isn't it?

---

In the other hand, the situation for logout is the same as @mightyraven . I fixed it temporarily by editing the `~/.xsession` file. Now my `.xsession` file is like this:

```sh
# cat ~/.xsession
umount ~/thinclient_drives
xfce4-session
```

It works well for remote drives when I logout and then reconnect.

**Comment by sh-dvl** [comment]
@metalefty:

after having the same problem, I've added xfuse_deinit() in main_cleanup() in file sesman/chansrv/chansrv.c and now all seem to work right.

**Comment by langemar** [comment]
@sh-dvl if you have a fix working for you could you create a PR for others to test and possible to merge it upstream if it works for all.

And thanks for your work :)

**Comment by metalefty** [comment]
@moobyfr can you confirm if the issue solved by #1140?

**Comment by moobyfr** [comment]
First test shows no problem, I switched a server at work to see the remaining mounts at end of the day.
